src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -31,11 +31,110 @@ public struct CallGraphNode
+        /// type parameters of the specified parent callable (the "innermost resolutions"). This mapping may potentially be to 
+        /// type parameters of other callables, which are then further concretized by subsequent resolutions. 
+        /// Returns the constructed dictionary as out parameter. Returns true if the combination of the given resolutions is valid, 
+        /// i.e. if there are no conflicting resolutions and type parameters of the parent callables are uniquely resolved 
+        /// to either a concrete type, a type parameter of another callable, or themselves.
+        /// Throws an ArgumentNullException if the given parent is null. 
+        /// Throws an ArgumentException if the given resolutions imply that type parameters from multiple callables are 
+        /// simultaneously treated as concrete types. 
+        /// NOTE: This routine prioritizes the verifications to ensure the correctness of the resolution over performance.  
+        /// </summary>
+        public static bool TryCombineTypeResolutions
+            (QsQualifiedName parent, out ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType> combined,
+            params ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>[] resolutions)
+        {
+            if (parent == null) throw new ArgumentNullException(nameof(parent));
+            var combinedBuilder = ImmutableDictionary.CreateBuilder<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>();
+            var success = true;
+
+            static Tuple<QsQualifiedName, NonNullable<string>> AsTypeResolutionKey(QsTypeParameter tp) => Tuple.Create(tp.Origin, tp.TypeName);
+            static bool ResolutionToTypeParameter(Tuple<QsQualifiedName, NonNullable<string>> typeParam, ResolvedType res) =>
+                res.Resolution is ResolvedTypeKind.TypeParameter tp && tp.Item.Origin.Equals(typeParam.Item1) && tp.Item.TypeName.Equals(typeParam.Item2);
+
+            foreach (var resolution in resolutions)
+            {
+                // Contains a lookup of all the keys in the combined resolutions whose value needs to be updated 
+                // if a certain type parameter is resolved by the currently processed dictionary. 
+                var mayBeReplaced = combinedBuilder
+                    .Where(kv => kv.Value.Resolution.IsTypeParameter)
+                    .ToLookup(
+                        kv => AsTypeResolutionKey(((ResolvedTypeKind.TypeParameter)kv.Value.Resolution).Item),
+                        entry => entry.Key);
+
+                // We need to ensure that the mappings for external type parameters are processed first, 
+                // to cover an edge case that would otherwise be indicated as a conflicting resolution.
+                var entriesToProcess = resolution.OrderBy(entry => entry.Key.Item1.Equals(parent));
```

Comment:
Reposting this comment as the original got marked as 'outdated'.

Because the looping logic also discriminates on this same predicate, it might be cleaner to just split this dictionary into two dictionaries based on this predicate and looping over these two separately, splitting out the looping logic appropriately.

Reply:
done